crates/ruff_linter/src/rules/flake8_trio/rules/unneeded_sleep.rs

```diff
@@ -0,0 +1,65 @@
+use ruff_diagnostics::{Diagnostic, Violation};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_ast::{self as ast, Expr, Stmt};
+use ruff_text_size::Ranged;
+
+use crate::checkers::ast::Checker;
+
+/// ## What it does
+/// Checks for the use of `trio.sleep` in a `while` loop.
+///
+/// ## Why is this bad?
+/// Instead of sleeping in a `while` loop, and waiting for a condition
+/// to become true, it's preferable to `wait()` on a `trio.Event`.
+///
+/// ## Example
+/// ```python
+/// DONE = False
+///
+/// async def func():
+///     while not DONE:
+///         await trio.sleep(1)
+/// ```
+///
+/// Use instead:
+/// ```python
+/// DONE = trio.Event()
+///
+/// async def func():
+///     await DONE.wait()
+/// ```
+#[violation]
+pub struct TrioUnneededSleep;
+
+impl Violation for TrioUnneededSleep {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("Use `trio.Event` instead of awaiting `trio.sleep` in a `while` loop")
+    }
+}
+
+pub(crate) fn unneeded_sleep(checker: &mut Checker, while_stmt: &ast::StmtWhile) {
```

Comment:
In general, we prefer passing in the typed node (`ast::StmtWhile` instead of the individual properties of it, like `body`). That way, users can't accidentally pass an AST node of the wrong type.